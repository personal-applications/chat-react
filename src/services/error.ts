import Ajv from "ajv";

const ajv = new Ajv({ removeAdditional: true });

export type ServerError = {
  status: string;
  code: number;
  message: string;
  errors: {
    field: string;
    message: string;
  }[];
};

const errorSchema = {
  type: "object",
  properties: {
    status: {
      type: "string",
    },
    code: {
      type: "number",
    },
    message: {
      type: "string",
    },
    errors: {
      type: "array",
      items: {
        type: "object",
        properties: {
          field: {
            type: "string",
          },
          message: {
            type: "string",
          },
        },
        required: ["field", "message"],
      },
    },
  },
  required: ["status", "code", "message", "errors"],
};

const validate = ajv.compile(errorSchema);
export const isServerFromError = (e: unknown) => {
  const valid = validate(e);
  return valid;
};
