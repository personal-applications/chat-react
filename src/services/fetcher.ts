import { JWT_KEY, getFromLocalStorage } from "../helpers/localStorage";

export async function getData<Data, Key extends string>(
  url: Key,
): Promise<Data> {
  try {
    const response = await fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${getFromLocalStorage(JWT_KEY)}`,
      },
    });

    if (!response.ok) {
      throw await response.json();
    }

    if (response.status === 204) {
      return {} as Data;
    }
    return await response.json();
  } catch (error) {
    console.error("Error fetching data:", error);
    throw error;
  }
}

export async function postData<Data, Key extends string, ExtraArgs>(
  url: Key,
  { arg }: { arg: ExtraArgs },
): Promise<Data> {
  try {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(arg),
    });

    if (!response.ok) {
      throw await response.json();
    }

    if (response.status === 204) {
      return {} as Data;
    }
    return await response.json();
  } catch (error) {
    console.error("Error fetching data:", error);
    throw error;
  }
}
