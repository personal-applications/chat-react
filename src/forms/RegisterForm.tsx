import { ajvResolver } from "@hookform/resolvers/ajv";
import { JSONSchemaType } from "ajv";
import { fullFormats } from "ajv-formats/dist/formats";
import { useForm } from "react-hook-form";
import { If, Then } from "react-if";
import Input from "../components/Input";
import PasswordInput from "../components/PasswordInput";
import { PASSWORD_REGEX } from "../constants";
import { ServerError, isServerFromError } from "../services/error";

export type RegisterFormFields = {
  email: string;
  password: string;
  confirmPassword: string;
  firstName: string;
  lastName: string;
};

const schema: JSONSchemaType<RegisterFormFields> = {
  type: "object",
  properties: {
    email: { type: "string", format: "email" },
    password: { type: "string", pattern: PASSWORD_REGEX },
    confirmPassword: { type: "string", pattern: PASSWORD_REGEX },
    firstName: { type: "string", minLength: 2 },
    lastName: { type: "string", minLength: 2 },
  },
  required: ["email", "password", "confirmPassword"],
  additionalProperties: false,
  errorMessage: {
    properties: {
      firstName: "First name must be at least 2 characters long.",
      lastName: "Last name must be at least 2 characters long.",
      email: "Invalid email format.",
      password:
        "Password must be at least 8 characters long and contain at least one uppercase letter, one lowercase letter, one number, and one special character.",
      confirmPassword: "Passwords do not match.",
    },
  },
};

export type Prop = {
  onSubmit: (data: RegisterFormFields) => Promise<void>;
};

export default function RegisterForm({ onSubmit }: Prop) {
  const {
    register,
    handleSubmit,
    formState: { errors },
    setError,
  } = useForm<RegisterFormFields>({
    resolver: ajvResolver(schema, { formats: fullFormats }),
    reValidateMode: "onChange",
  });

  const asyncOnSubmit = async (data: RegisterFormFields) => {
    try {
      await onSubmit(data);
    } catch (error) {
      if (isServerFromError(error)) {
        const serverError = error as ServerError;
        if (serverError.errors.length) {
          serverError.errors.forEach((e) => {
            setError(e.field as keyof RegisterFormFields, {
              message: e.message,
            });
          });
        } else {
          setError("root", { message: serverError.message });
        }
      } else {
        setError("root", { message: "An unknown error occurred" });
      }
    }
  };

  return (
    <form
      noValidate
      onSubmit={handleSubmit(asyncOnSubmit)}
      className="mt-6 flex flex-col gap-y-6 rounded-md bg-base-100 p-8"
    >
      <If condition={!!errors.root}>
        <Then>
          <div role="alert" className="alert alert-error">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              className="stroke-current shrink-0 h-6 w-6"
              fill="none"
              viewBox="0 0 24 24"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                d="M10 14l2-2m0 0l2-2m-2 2l-2-2m2 2l2 2m7-2a9 9 0 11-18 0 9 9 0 0118 0z"
              />
            </svg>
            <span>{errors.root?.message}</span>
          </div>
        </Then>
      </If>
      <div className="flex gap-x-6">
        <Input
          type="text"
          placeholder="First name"
          error={errors.firstName?.message}
          {...register("firstName")}
        />
        <Input
          type="text"
          placeholder="Last name"
          error={errors.lastName?.message}
          {...register("lastName")}
        />
      </div>
      <Input
        type="email"
        placeholder="Email"
        error={errors.email?.message}
        {...register("email")}
      />
      <PasswordInput
        placeholder="Password"
        error={errors.password?.message}
        {...register("password")}
      />
      <PasswordInput
        placeholder="Confirm password"
        error={errors.confirmPassword?.message}
        {...register("confirmPassword")}
      />

      <button className="btn btn-primary" type="submit">
        Sign up
      </button>
    </form>
  );
}
