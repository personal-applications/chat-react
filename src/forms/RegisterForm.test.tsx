import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ServerError } from "../services/error";
import RegisterForm, { RegisterFormFields } from "./RegisterForm";

test("renders correctly", () => {
  const tree = render(<RegisterForm onSubmit={vi.fn()} />);

  expect(tree).toMatchSnapshot();
});

test("displays validation errors when fields are left empty", async () => {
  const handleSubmit = vi.fn();
  render(<RegisterForm onSubmit={handleSubmit} />);

  await userEvent.click(screen.getByText(/Sign up/));

  expect(screen.getAllByText("Value is too short.")).toHaveLength(2);
  expect(screen.getByText("Invalid email address.")).toBeInTheDocument();
  expect(screen.getAllByText("Invalid value.")).toHaveLength(2);
});

test("submits form data", async () => {
  const handleSubmit = vi.fn();
  render(<RegisterForm onSubmit={handleSubmit} />);

  await userEvent.type(screen.getByPlaceholderText("First name"), "John");
  await userEvent.type(screen.getByPlaceholderText("Last name"), "Doe");
  await userEvent.type(
    screen.getByPlaceholderText("Email"),
    "john.doe@example.com",
  );
  await userEvent.type(screen.getByPlaceholderText("Password"), "Aa123456!");
  await userEvent.type(
    screen.getByPlaceholderText("Confirm password"),
    "Aa123456!",
  );
  await userEvent.click(screen.getByText(/Sign up/));

  expect(handleSubmit).toHaveBeenCalledWith({
    firstName: "John",
    lastName: "Doe",
    email: "john.doe@example.com",
    password: "Aa123456!",
    confirmPassword: "Aa123456!",
  });
});

test("displays server error message when server returns field-specific error", async () => {
  const serverError: ServerError = {
    code: 400,
    message: "Bad Request",
    status: "error",
    errors: [
      {
        field: "email",
        message: "Email is already in use",
      },
    ],
  };
  const handleSubmit = vi.fn().mockRejectedValue(serverError);
  render(<RegisterForm onSubmit={handleSubmit} />);

  const formData: RegisterFormFields = {
    firstName: "John",
    lastName: "Doe",
    email: "john.doe@example.com",
    password: "Aa123456!",
    confirmPassword: "Aa123456!",
  };

  await userEvent.type(
    screen.getByPlaceholderText("First name"),
    formData.firstName,
  );
  await userEvent.type(
    screen.getByPlaceholderText("Last name"),
    formData.lastName,
  );
  await userEvent.type(screen.getByPlaceholderText("Email"), formData.email);
  await userEvent.type(
    screen.getByPlaceholderText("Password"),
    formData.password,
  );
  await userEvent.type(
    screen.getByPlaceholderText("Confirm password"),
    formData.confirmPassword,
  );

  await userEvent.click(screen.getByText(/Sign up/));

  expect(handleSubmit).toHaveBeenCalled();
  expect(screen.getByText("Email is already in use")).toBeInTheDocument();
});

test("display unknown error message when something goes wrong", async () => {
  const serverError = {
    message: "Something goes wrong",
  };
  const handleSubmit = vi.fn().mockRejectedValue(serverError);
  render(<RegisterForm onSubmit={handleSubmit} />);

  const formData: RegisterFormFields = {
    firstName: "John",
    lastName: "Doe",
    email: "john.doe@example.com",
    password: "Aa123456!",
    confirmPassword: "Aa123456!",
  };

  await userEvent.type(
    screen.getByPlaceholderText("First name"),
    formData.firstName,
  );
  await userEvent.type(
    screen.getByPlaceholderText("Last name"),
    formData.lastName,
  );
  await userEvent.type(screen.getByPlaceholderText("Email"), formData.email);
  await userEvent.type(
    screen.getByPlaceholderText("Password"),
    formData.password,
  );
  await userEvent.type(
    screen.getByPlaceholderText("Confirm password"),
    formData.confirmPassword,
  );

  await userEvent.click(screen.getByText(/Sign up/));

  expect(handleSubmit).toHaveBeenCalled();
  expect(screen.getByText("An unknown error occurred")).toBeInTheDocument();
});
