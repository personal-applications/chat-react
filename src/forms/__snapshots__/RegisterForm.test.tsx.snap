// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="mt-6 flex flex-col gap-y-6 rounded-md bg-base-100 p-8"
        novalidate=""
      >
        <div
          class="flex gap-x-6"
        >
          <div
            class="flex w-full flex-col"
          >
            <label
              class="input input-bordered flex items-center gap-2"
            >
              <input
                class="grow"
                name="firstName"
                placeholder="First name"
                type="text"
              />
            </label>
          </div>
          <div
            class="flex w-full flex-col"
          >
            <label
              class="input input-bordered flex items-center gap-2"
            >
              <input
                class="grow"
                name="lastName"
                placeholder="Last name"
                type="text"
              />
            </label>
          </div>
        </div>
        <div
          class="flex w-full flex-col"
        >
          <label
            class="input input-bordered flex items-center gap-2"
          >
            <input
              class="grow"
              name="email"
              placeholder="Email"
              type="email"
            />
          </label>
        </div>
        <div
          class="flex w-full flex-col"
        >
          <label
            class="input input-bordered flex items-center gap-2"
          >
            <input
              class="grow"
              name="password"
              placeholder="Password"
              type="password"
            />
            <button
              type="button"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M396 512a112 112 0 10224 0 112 112 0 10-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
                />
              </svg>
            </button>
          </label>
        </div>
        <div
          class="flex w-full flex-col"
        >
          <label
            class="input input-bordered flex items-center gap-2"
          >
            <input
              class="grow"
              name="confirmPassword"
              placeholder="Confirm password"
              type="password"
            />
            <button
              type="button"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M396 512a112 112 0 10224 0 112 112 0 10-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
                />
              </svg>
            </button>
          </label>
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Sign up
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="mt-6 flex flex-col gap-y-6 rounded-md bg-base-100 p-8"
      novalidate=""
    >
      <div
        class="flex gap-x-6"
      >
        <div
          class="flex w-full flex-col"
        >
          <label
            class="input input-bordered flex items-center gap-2"
          >
            <input
              class="grow"
              name="firstName"
              placeholder="First name"
              type="text"
            />
          </label>
        </div>
        <div
          class="flex w-full flex-col"
        >
          <label
            class="input input-bordered flex items-center gap-2"
          >
            <input
              class="grow"
              name="lastName"
              placeholder="Last name"
              type="text"
            />
          </label>
        </div>
      </div>
      <div
        class="flex w-full flex-col"
      >
        <label
          class="input input-bordered flex items-center gap-2"
        >
          <input
            class="grow"
            name="email"
            placeholder="Email"
            type="email"
          />
        </label>
      </div>
      <div
        class="flex w-full flex-col"
      >
        <label
          class="input input-bordered flex items-center gap-2"
        >
          <input
            class="grow"
            name="password"
            placeholder="Password"
            type="password"
          />
          <button
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M396 512a112 112 0 10224 0 112 112 0 10-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
              />
            </svg>
          </button>
        </label>
      </div>
      <div
        class="flex w-full flex-col"
      >
        <label
          class="input input-bordered flex items-center gap-2"
        >
          <input
            class="grow"
            name="confirmPassword"
            placeholder="Confirm password"
            type="password"
          />
          <button
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M396 512a112 112 0 10224 0 112 112 0 10-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
              />
            </svg>
          </button>
        </label>
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Sign up
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
